Issue 44682: Run PX data validation as a pipeline job

vagisha opened the pull request:
New PR after renaming branch to target release22.3-SNAPSHOT.  Original PR: https://github.com/LabKey/MacCossLabModules/pull/175

#### Rationale
We would like to run the data validation step of submitting a dataset to Panorama Public as a pipeline job.  The details are described in this doc: https://docs.google.com/document/d/1zDJY5f8rbb1Nq95__JxFStVmPqtcufFkVV3FY_wNNw0/edit

#### Related Pull Requests
* https://github.com/LabKey/targetedms/pull/497

#### Changes
- Run data validation for ProteomeXchange as a pipeline job
- Results from running the job are stored in new tables added to the panoramapublic schema
- PxValidationStatusApiAction returns the validation results as JSON which is displayed in pxValidationStatus.jsp
- CopyExperimentAction, PX actions (validate / submit / update PX XML), and PX XML and HTML writers have been updated to use the new data validation classes
- A spectral library is marked as incomplete if:
  - the SpectrumSourceFiles table of the .blib does not have the idFileName column, or there are now rows in the SpectrumSourceFiles table
  - the .blib library was built with an assay library (.csv files as source)
  - library is unsupported (only .blib and .elib are supported)
  - the library is referenced in the Skyline XML but not included in the .sky.zip
- For .blib built with MaxQuant search results, look for mqpar.xml etc. in addition to what is listed in the SpectrumSourceFiles.idFileName column
- If submitted data is in sub-folders we look for spectral library source files in the parent experiment folder as well as the sub-folders that have a Skyline document with the library. The files have to be uploaded to only one folder
- Sample files in a Skyline document that have the same file name but were imported from different paths will be marked as ambiguous, and the dataset will not be considered a valid PX dataset.  We may support this later by requiring that such files be uploaded into subfolders named for the replicate so that we can can link to the correct sample file.
- Added some fixes to matching modifications to Unimod in the case that a modification in the Skyline document does not have a Unimod ID. Updated test in ExperimentModificationGetter with modifications found on PanoramaWeb without a Unimod Id.
-  ExperimentUserForeignKey extends UserIdForeignKey so that the Submitter and LabHead columns can include all users when the experiment list is viewed in the Panorama Public project.
- Updated existing tests and added a PanoramaPublicValidationTest
- Moved test sample data to TargetedMS/panoramapublic subfolder

vagisha committed:
Issue 44682: Run PX data validation as a pipeline job
- Run data validation for ProteomeXchange as a pipeline job
- Results from running the job are stored in new tables added to the panoramapublic schema
- PxValidationStatusApiAction returns the validation results as JSON which is displayed in pxValidationStatus.jsp
- Updated CopyExperimentAction, PX actions (validate / submit / update PX XML), and PX XML and HTML writers to use the new data validation classes
- Spectral libraries are marked as incomplete if:
-- libraries that are not included in the sky.zip but are referenced in the Skyline XML
-- old libraries without an idFileName column in the SpectrumSourceFiles
-- libraries built with assay libraries (.csv files as source)
-- unsupported libraries: anything other than .blib or .elib
- For .blib built with MaxQuant search results, look for mqpar.xml etc. in addition to what is listed in the SpectrumSourceFiles table
- If submitted data is in sub-folders we look for spectral library source files in the parent experiment folder as well as the sub-folders that have a Skyline document with the library
- Added some fixes to matching modifications to Unimod in case the modification in the Skyline document does not have a Unimod ID. Updated test in ExperimentModificationGetter with several modifications found on PanoramaWeb without a Unimod Id.
- Sample files in a Skyline document that have the same file name but were imported from different paths will be marked as ambiguous
- Updated existing tests and added a PanoramaPublicValidationTest
- Moved test sample data to TargetedMS/panoramapublic subfolder
- In SpecLibReader.readLibSourceFiles() return empty list instead of throwing exception if the library file size is 0-bytes.
- ExperimentUserForeignKey extends UserIdForeignKey so that the Submitter and LabHead columns can include all users when the experiment list is viewed in the Panorama Public project.

vagisha committed:
Fix for failing PanoramaPublicValidationTest.

vagisha committed:
Expand rows as needed in DataValidationPage. Display '-' instead of '0' in the size column for libraries missing in the sky.zip.

vagisha committed:
- Code review changes
-- PanoramaPublicTable overrides supportsContainerFilter(); returns false if noContainerFilter is set.
-- Added constants for statusId in pxValidationStatus.jsp
- Bug fixes
-- Populate form with value of incompletePxSubmission when doing a resubmit or update
-- Use the correct container with AuditLogService.get().getAuditEvents()

vagisha committed:
Merge remote-tracking branch 'remotes/origin/release22.3-SNAPSHOT' into 22.3_fb_44682_data-validation-pipeline-job

vagisha committed:
Merge remote-tracking branch 'remotes/origin/release22.3-SNAPSHOT' into 22.3_fb_44682_data-validation-pipeline-job

labkey-jeckels commented on panoramapublic/resources/schemas/dbscripts/postgresql/panoramapublic-22.000-22.001.sql:
```diff
@@ -0,0 +1,141 @@
+    Name                       VARCHAR(300) NOT NULL, -- Name of the Skyline file
+
+    CONSTRAINT PK_SkylineDocValidation PRIMARY KEY (Id),
+    CONSTRAINT FK_SkylineDocValidation_DataValidation FOREIGN KEY (ValidationId) REFERENCES panoramapublic.DataValidation(Id)
```
Should probably have a FK on container too

vagisha replied:
I removed the container column.  Getting the container from the targetedms run instead.

vagisha committed:
Modification information interface (#188)

* - Added an interface for users to:
  -- Find a Unimod match for modifications that did not have a Unimod Id in the Skyline document .
  -- Define a combination modification for a structural modification that is a combination of two modifications.
- Removed automatic Unimod matching from the ProteomeXchange DataValidator.
- Updated PX validation details UI to show user-assigned Unimod Ids and combination modifications.
- PX validation status is updated when a user-assigned modification information is added or deleted
- Modification information saved in the source folder is copied to the Panorama Public copy.
- Added Formula and ChemElement classes to build / add / subtract and normalize modification formulas
- Added test coverage

vagisha committed:
Modification information interface pt2 (#195)

- Updated unimod_NO_NAMESPACE.xml
- Unimod modifications that have a heavy isotope (2H, 13C, 15N, 18O) will be considered an isotopic modification
- Any isotopic modification that is also a chemical modification (i.e. it changes the elemental composition of the residue or terminus) will also be considered a structural modification
- Updated the PX validator to automatically assign Unimod Ids to hard-coded modifications in Skyline
- Updated the PX validator to automatically assign Unimod Ids to wildcard modifications based on the modified amino acids in the peptides in the Skyline documents
-- Added IsotopeUnimodInfo table to support this
-- Wildcard modification matches are recalculated for each data validation job because the modified sites may have changed due to Skyline doc addition / deletion
- For isotope labels of the form Dimethyl:, mTRAQ:, ICAT-C: etc., a diff isotope formula is calculated (diff of modification minus the parent structural modification). Example: diff isotope formula for  mTRAQ:13C(6)15N(2) is  mTRAQ:13C(6)15N(2) minus mTRAQ, which is H12CC'6N'2O - H12C7N2O = C'6N'2 - C6N2. This is how isotope formulas are setup in Skyline. For isotopic formulas a match is computed against the formula of the Unimod modification as well as the diff isotope formula.
- Keep UnimodModifications in a BlockingCache for 1 day.
- Added queries for viewing the HTML-formatted modified peptide / precursor sequence along with modification name, modified index and mass diff.
- Throw UnexpectedException if Unimod.xml could not be parsed
- Updated PXHtml and PxXml writers to write multiple modification matches for wildcard modifications and combination modifications.
- Add ExperimentModInfo to Modification of _modInfoId is not null.
- ExperimentModificationGetter.getStructuralUnimodMod() and getIsotopicUnimodMod() throw IllegalArgumentException if the formula could not be parsed or not calculated.
- Updated test and added new test documents.
- Updated test data file with heavyK_R modification on amino acids that don't have the same number of labeled atoms.

vagisha committed:
- Removed container column from SkylineDocValidation table. Get container from the targetedms run instead, if it exists
- If the run no longer exists (e.g. if it was deleted, or if the subfolder it was in was deleted), reports for older validation jobs that were run prior to the deletion will not display links to the document or its peptides.

vagisha committed:
- Use acquired time in addition to the file path imported into the Skyline document when looking for ambiguous sample files
- Look at sample files in all the documents within a container when marking files as ambiguous.
- If a sample file (or library source file) is found on the server, display the path relative to the container file root
- Display folder path (relative to experiment container) for Skyline documents if the experiment contains sub-folders.
- Added a SampleFileId column to SkylineDocSampleFile table
- Added table headers to sample files table and spectral library source files tables in pxValidationStatus.jsp
- Added query and default query view to view sample files with the replicate name, skyline document, instrument, serial number and sample acquisition time
- For ambiguous sample files add a link to view sample files with the same name in the folder that contains the Skyline document
- Added test and new test documents for testing ambiguous sample files.

vagisha committed:
- Make validation summary panel cleaner.
- Change order of problems displayed in the validation summary.
- Change order and content of text displayed when a validation job is outdated.
- CSS changes, and htmlEncode.
- Added "Upload" column to the spectral library grid if library source files are missing.
- Show PX validation summary higher up on the experiment details page.
- Display experiment title in the validation summary panel only if we are not on the experiment details page, which already shows the title.
- Show the find match link in the modifications table only if user has permissions to make edits.
- SpecLibView displayed on the experiment details page cannot be customized (similar to the modifications tables). The grid header links to the query where the view can be customized.
- User can override the default and 'SpectralLibrariesInfo' views on the SpectralLibraries query.
- Updates to log / error messages.

vagisha committed:
- Added a "Log" column (warps the JobId column) to DataValidationTableInfo. We don't want to display the job status in this table. A "COMPLETE" status could be confusing.
- Add custom errors view where back button links to the returnURL in the form, if present. Otherwise the button text is "View Experiment Details" and it links to the experiment details page.
- Replace instances of "pipeline job log" with "validation log".
- Do not display validation details (date run etc.) on the experiment details page if the validation is outdated. Show only a "..validation is outdated..." in this case
- Display the "..validation is outdated.." message at the top of validation details page
- Display a note about ambiguous sample files

vagisha committed:
.wiff2 files need .wiff.scan not .wiff2.scan

vagisha committed:
- Link user-entered information for spectral libraries with spectral library data validation. This can affect the validation status of a library. For example, if the dependency on the library is "Irrelevant to results" then we will not require the source files used to build the library. A specLibInfoId column was added to the SpecLibValidation table.
- SpecLibView displayed on the experiment details page cannot be customized (similar to the modification info tables). The grid header links to the query where the view can be customized.
- Added an action class to delete a row in the SpecLibInfo table. Added link to this action in the "Library Info" column.
- Data validation status is no longer updated automatically when a Unimod match is saved or deleted. Instead, the validation job will be considered "outdated" and it will have to be run again.
- After adding a Unimod match, the user can go to a page (ViewExperimentModifications action class) that shows all the structural and isotope modifications in the experiment and associate any other modifications with Unimod Ids, if required.
- A data validation job is considered "outdated" if any new modification info rows or spectral library info rows were added or deleted after the validation job's create date. This is in addition to checking that the runs that were validated are the same ones that are in the folder, and checking that there are no file system events in the audit log after the validation job started.
- The "inferred" column on the ModificationValidation table is now only set if a new mod info is created during the data validation job (e.g. for wild-card and hard-coded Skyline modifications). This is so that we can ignore these modifications when we determine if a validation is outdated.  Otherwise, the validation would be considered outdated right after it is run because the mod infos created during the validation job would have a newer create date than the data validation row. For the validation details page, use the isMatchAssigned() method instead to mark those modifications that were assigned a Unimod Id.
- Added test coverage and new test file

vagisha committed:
fix for Prosit libraries.

labkey-jeckels commented on panoramapublic/src/org/labkey/panoramapublic/view/publish/pxValidationStatus.jsp:
```diff
@@ -638,7 +636,7 @@
             if (hasInferred) {
                 var noteHtml = "Modification names starting with <strong>**</strong> in the Name column did not have a Unimod Id in the Skyline document."
-                        + " A Unimod match was was inferred based on the formula, modification site(s) and terminus in the modification definition"
+                        + " A Unimod match was was assigned based on the formula, modification site(s) and terminus in the modification definition"
```
Typo - "was was"

vagisha committed:
- Changed order of columns in the ExperimentSampleFile query view.
- Do not sort ExperimentIsotopeModInfo and ExperimentStructuralModInfo table views on the UnimodMatch column.
- Changed text for outdated data validation message.
- Fixed UnimodModification.getLink() to remove margin on link.
- Updated display of the UnimodMatch column to be similar to the modification results grid when a Unimod match is missing.
- Added button to start validation job on the validation results page if validation results are outdated.
- Removed "Upload" column in the "Skyline Document Sample Files" grid on the validation results page.  An upload button was added at the top of the sample files table in the nested view.
- Updated ambiguous sample files message.
- Moved ambiguous files link in the "Status" column.

vagisha committed:
- Add message about assigned Unimod matches on ViewExperimentModifications
- Add the '**' next to assigned Unimod Id instead of Unimod name to be consistent with the modifications grids
- Renamed "Unimod Id" column on the validation results page to "Unimod Match" to be consistent with the modifications grids
- Sort Skyline documents by container and RunId; Sort modifications by mod type and modification Id

vagisha committed:
Added link to navigate to the source experiment folder if we are looking at a Panorama Public copy.

vagisha committed:
Only site admins can see the list of data validation jobs and delete them. Other users only see the latest validation results.

vagisha committed:
- Separated pxValidationStatus.jsp into two.  PxValidationStatusAction returns pxValidationRunning.jsp if the job is running, and pxValidationStatus.jsp if the job is complete.
- Updated tests.

vagisha committed:
-Validation status gets updated automatically if a modification information is added / deleted OR if spectral library info is added / edited / deleted.
- set returnUrl when linking to these actions from the validation results page.

vagisha committed:
Added link to delete a mod info from the validation results page.

vagisha committed:
- Added additional information to the SpecLib JSON
- Updated library validation status messages
- Display assigned Unimod matches and delete link in a single line
- Moved note about inferred modification matches to a docked item.
- Sort libraries by id
- Removed upload column from spectral libraries validation grid. Added upload files button inside the expanded row instead
- Updates to library status and library info rendering
- Changed default return URL from saving / deleting a Unimod match to be ViewExperimentModifications

vagisha committed:
Fixed PanoramaPublicModificationsTest.

vagisha committed:
"Skip Subfolders and Continue" should not do a POST request. Updated test.

vagisha committed:
Fixed return and cancel URLs for saving / deleting modification information.

vagisha committed:
- Do not limit source type values for unsupported libraries (e.g. NIST)
- Bug fix in DataValidationManager.getLibrariesMatchingSpecLibInfo() - do not use the experiment annotations container when getting an ISpectrumLibrary. The Skyline document (and library) may be in a subfolder.
- If source accession in SpecLibInfoTableInfo is a PXD, render a link to the PXD on ProteomeCentral

vagisha committed:
Fixed tests.

labkey-jeckels commented on panoramapublic/src/org/labkey/panoramapublic/query/DataValidationManager.java:
```diff
@@ -396,6 +394,182 @@ public static DataValidation updateDataValidation(DataValidation validation, Use
+        List<SpecLib> specLibs = new TableSelector(PanoramaPublicManager.getTableInfoSpecLibValidation(), filter, null).getArrayList(SpecLib.class);
+        if (specLibs.size() > 0)
+        {
+            specLibList.addAll(specLibs);
```
Is this needed? Could you just return specLibs directly and avoid the copy?

vagisha committed:
- Added action classes to display a "Upload Files" page that has a FWP along with a grid showing Skyline documents with missing sample files OR spectral libraries with missing source files.  The file root of the FWP is set to @files/RawFiles. This page also has a button to re-run validation.
- "Upload Files" buttons on the data validation page link to the new actions instead of going to the "Raw Data" tab.
- Moved CSS and the code to render the sample files and spectral libraries grids from pxValidationStatus.jsp to to pxValidation.css and pxValidation.js

vagisha committed:
- Changed order of columns in ExperimentSampleFiles query view.
- Separate help messages for library status and library info.
- Updates text displayed for library status and library info in the validation results page.
- Removed unused code in DataValidationManager that determined if a validation is outdated when mod info or spec lib info are added / modified.
- "Library Info" column is blank if user does not have edit permission.
- Added a ProteomeXchange validation status indicator in the "Targeted MS Experiment" webpart.
- Display "No modifications found" as a docked item in the modifications validation grid, similar to the spectral libraries validation grid.
- Prevent expand on double click in the validation grids.

vagisha committed:
- Export all module specific properties in ExperimentExportTask. This had been commented out when we had different values for properties such as SKIP_CHROMATOGRAM_IMPORT_PROPERTY in user projects, and we wanted the copied folders to inherit the values in the Panorama Public project.
- Renamed UpdateFolderTypeTask -> CopyLibraryStateTask since it now only copies the library state for chromatogram library folders.
- Added 'QC' folder to folder types supported for copying to Panorama Public.

vagisha committed:
- Update message text when validation is outdated.
- Do not check for "modifications.xml". It is not always used to build a library from MaxQuant results.
- Display library info when looking at the "Upload missing files" page.

vagisha committed:
Use LabKey SQL's lookup syntax instead of explicit JOINs for more efficient SQL.

vagisha committed:
bug fix in check for libraries built with .csv files.

vagisha committed:
Fix unit test since we are no longer requiring modifications.xml.

vagisha committed:
Merge remote-tracking branch 'remotes/origin/release22.3-SNAPSHOT' into 22.3_fb_44682_data-validation-pipeline-job
